src/Servers/Kestrel/Core/src/Internal/Http/ChunkWriter.cs

```diff
@@ -44,6 +37,33 @@ public static int BeginChunkBytes(int dataCount, Span<byte> span)
             return count;
         }
 
+        internal static int GetPrefixLength(Memory<byte> memory, int suffixLength)
+        {
+            var prefixLength = GetBeginChunkByteCount(memory.Length);
+
+            // Super edge case. If we call GetMemory and it returns 4099, we would originally calculate _currentMemoryPrefixBytes
+            // as 6 bytes. However, after subtracting the bytes for _currentChunkMemory.Length and the endPrefix, the real
+            // body length would be 5 bytes. 
+            return GetBeginChunkByteCount(memory.Length - prefixLength - suffixLength);
```

Comment:
This logic isn't stable on the boundaries. This [gist](https://gist.github.com/halter73/af2b9f78978f83813b19e187c4e5309e) demonstrates how shortening the expected prefix length one byte with the second call to GetBeginChunkByteCount() allows the data length to become one byte longer once again necessitating a longer prefix.

> Given 21 bytes from underlying PipeWriter, slice 1 extra byte off end of actualMemory.

The above output from the gist shows that Memory with a length of 21 bytes is one of the problematic inputs. With this logic, prefixLength is 4 after the first line since it takes 2 hex chars to represent 21 (0x15) plus CRLF. The method the turns around and returns 3 though because 15 (21 - 4 - 2) can be represented with 1 hex char (0xF).

I think ultimately GetPrefixLength should return the lower number in these boundary cases, because returning the larger prefix length would always result in shifting data to the left in Advance() even when the buffer is completely filled. If we return the lower number though, we have to slice one extra byte off the end of the Memory Http1OutputProducer returns from GetMemory() so we don't end up in the situation where we need the longer prefix.

I'm wondering if we should just hard code the mapping from memory lengths to prefix lengths. With 8 or 16 return values depending on whether we special-case the boundaries, it should only take 3 or 4 comparisons. Or we could create a benchmark and let twitter go at it.

and I'm wondering if we should just hard code the ranges on this logic.

Reply:
I reached the same conclusion as you too. I have a slightly different approach though, I’ll explain tomorrow.